sig-api-machinery/design-docs/admission-control-webhooks.md

```diff
@@ -0,0 +1,934 @@
+
+We will clearly document the reasons for each and how users should decide which to use.
+
+
+## Monitorability 
+
+There should be prometheus variables to show:
+
+
+
+*   API operation latency
+    *   Overall
+    *   By webhook name
+*   API response codes
+    *   Overall
+    *   By webhook name.
+
+Adding a webhook dynamically adds a key to a map-valued prometheus metric. Webhook host process authors should consider how to make their webhook host monitorable: while eventually we hope to offer a set of best practices around this, for the initial release we won't have requirements here.
+
+
+## API Changes 
+
+GenericAdmissionWebhook Admission Controller is split and renamed.
+
+
+
+*   One is called `MutatingAdmissionWebhook`
+*   The other is called `ValidatingAdmissionWebhook`
+*   Splitting them allows them to appear in different places in the `--admission-control` flag's order.
+
+ExternalAdmissionHookConfiguration API is split and renamed.
+
+
+
+*   One is called  `MutatingAdmissionWebhookConfiguration`
+*   The other is called `ValidatingAdmissionWebhookConfiguration`
+*   Splitting them:
+    *   makes it clear what the order is when some items don't have both flavors,
+    *   enforces mutate-before-validate,
+    *   better allows declarative update of the config than one big list with an implied partition point
+
+The `ValidatingAdmissionWebhookConfiguration` stays the same as `ExternalAdmissionHookConfiguration` except it moves to v1beta1.
+
+The `MutatingAdmissionWebhookConfiguration` is the same API as `ValidatingAdmissionWebhookConfiguration`.  It is only visible via the v1beta1 version.
+
+The Group/Version called 
+
+`admissionregistration.k8s.io/v1alpha1` with kinds
+
+InitializerConfiguration and ExternalAdmissionHookConfiguration.
+
+InitializerConfiguration will not join `admissionregistration.k8s.io/v1beta1` at this time.
+
+Any webhooks that register with v1alpha1 may or may not be surprised when they start getting Versioned data.  But we don't make any promises for Alpha, and detecting which API was used to register is too complex.
+
+
+## Order of Admission 
+
+The recommended order on kubernetes.io for `--admission-control` will be: `MutatingAdmissionWebhook,NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,DefaultTolerationSeconds,ValidatingAdmissionWebhook,ResourceQuota`
+
+ This order has the following properties:
+
+
+
+*   ResourceQuota comes last, so that if prior ones reject a request, it won't increment quota.
+*   All other Static ones are in the order recommended by [the docs](https://kubernetes.io/docs/admin/admission-controllers/#is-there-a-recommended-set-of-plug-ins-to-use).  (which variously do mutation and validation) Preserves the behavior when there are no webhooks.
```

Comment:
I think some information needs to flow the static mutators to the dynamic mutators.  Things like the persistent volume labels provide information that is otherwise inaccessible.

Reply:
I can likewise imagine writing a dynamic webhook that modifies a field used as input by a static mutator. There may not be a single correct order that works for everyone.